components/sequencer/snowflake/snowflake_sequencer.go

```diff
@@ -0,0 +1,200 @@
+	mu         sync.Mutex
+	db         *sql.DB
+	smap       map[string]chan int64
+	biggerThan map[string]int64
+	logger     log.ErrorLogger
+	ctx        context.Context
+	cancel     context.CancelFunc
+}
+
+func NewSnowFlakeSequencer(logger log.ErrorLogger) *SnowFlakeSequencer {
+	return &SnowFlakeSequencer{
+		logger: logger,
+	}
+}
+
+func (s *SnowFlakeSequencer) Init(config sequencer.Configuration) error {
+	var err error
+	s.metadata, err = utils.ParseSnowflakeMetadata(config.Properties)
+	if err != nil {
+		return err
+	}
+	//for unit test
+	s.metadata.MysqlMetadata.Db = s.db
+
+	s.smap = make(map[string]chan int64)
+	s.biggerThan = config.BiggerThan
+	s.ctx, s.cancel = context.WithCancel(context.Background())
+
+	if s.workerId, err = utils.NewMysqlClient(&s.metadata.MysqlMetadata); err != nil {
+		return err
+	}
+	return err
+}
+
+func (s *SnowFlakeSequencer) GetNextId(req *sequencer.GetNextIdRequest) (*sequencer.GetNextIdResponse, error) {
+	s.mu.Lock()
+	var ok bool
+	seed, ok := s.smap[req.Key]
+	//If the key appears for the first time, start a new goroutine for it. If the key doesn't appear for a long time, close the goroutine
+	if !ok {
+		seed = make(chan int64, 1000)
+		s.smap[req.Key] = seed
+
+		var oldWorkerId int64
+		var oldTimeStamp int64
+		currentTimeStamp := time.Now().Unix() - s.metadata.StartTime
+		cts := currentTimeStamp
+
+		err := s.metadata.MysqlMetadata.Db.QueryRow("SELECT WORKER_ID, TIMESTAMP FROM "+s.metadata.MysqlMetadata.KeyTableName+" WHERE SEQUENCER_KEY = ?", req.Key).Scan(&oldWorkerId, &oldTimeStamp)
+		if err == nil {
+			if oldWorkerId == s.workerId {
+				cts = oldTimeStamp + 1
+			}
+		}
+
+		var sequence int64
+		var workerId = s.workerId
+		startId := cts<<s.metadata.TimestampShift | workerId<<s.metadata.WorkidShift | sequence
+
+		//producer
+		go s.producer(startId, currentTimeStamp, seed, req.Key)
+	}
+	s.mu.Unlock()
+
+	var id int64
+	for {
+		select {
+		case id, ok = <-seed:
+			if !ok {
+				return nil, errors.New("please try again or adjust the start time")
+			}
+			return &sequencer.GetNextIdResponse{
+				NextId: id,
+			}, nil
+		case <-time.After(s.metadata.ReqTimeout):
+			return nil, errors.New("request id time out")
+		}
+	}
+}
+
+func (s *SnowFlakeSequencer) GetSegment(req *sequencer.GetSegmentRequest) (support bool, result *sequencer.GetSegmentResponse, err error) {
+	return false, nil, nil
+}
+
+func (s *SnowFlakeSequencer) producer(id, currentTimeStamp int64, ch chan int64, key string) {
+	defer func() {
+		if x := recover(); x != nil {
+			log.DefaultLogger.Errorf("panic when producing id with snowflake algorithm: %v", x)
+		}
+	}()
+
+	var maxTimeStamp int64
+	var maxSeqId int64
+	var seq int64
+
+	maxTimeStamp = 2 << s.metadata.TimeBits
+	maxSeqId = 2 << s.metadata.SeqBits
+	for {
+		select {
+		case <-s.ctx.Done():
+			close(ch)
+			return
+		//if timeout, remove key from map and record key, workerId, timestamp to mysql
+		case <-time.After(s.metadata.KeyTimeout):
```

Comment:
time.After do not use In the for loop, this will new many many timer, put this outside for loop

Reply:
got it